Compute union for [EntityType].

bgogul opened the pull request:
This is a part of series of PRs for implementing type inference for `Recipe` in Kotlin.

csilvestrini commented on java/arcs/core/analysis/EntityTypeUnion.kt:
```diff
@@ -0,0 +1,70 @@
+ * Computes the union of the two [Schema] instances. Returns [Outcome.Failure] if the union
+ * is not possible as the inputs are incompatible.
+ */
+fun Schema.union(other: Schema): Outcome<Schema> {
```
This would be a great candidate for using `infix fun`. Then you can call it like so: `schema1 union schema2`

bgogul replied:
Good idea. Done.

csilvestrini commented on java/arcs/core/analysis/EntityTypeUnion.kt:
```diff
@@ -0,0 +1,70 @@
+ */
+private fun SchemaFields.union(other: SchemaFields): Outcome<SchemaFields> {
+    val newSingletons = singletons.unionFields(other.singletons)
+        .getOrElse { return Outcome.Failure(it) }
```
It occurs to me now that this `getOrElse` function is not that useful in its current form. A new helper function for this would be nice. e.g. `unwrapOr` (or you could reuse `getOrElse`, but change its lambda to accept the `Outcome.Failure` rather than just the failure message.

I'm imagining something like this:

```
fun Outcome.unwrapOr(block: (Outcome.Failure) -> Unit): T = when(this) {
  is Success -> value
  is Failure -> block(this)
}
```

Then you could use it like so:

```
a.union(b).unwrapOr { return it }
```

Which is a bit shorter.

(The typing of the lambda lets you easily get at the error message, if you still want to do that: `x.unwrapOr { it.reason }`)

bgogul replied:
I was also not happy with the `getOrElse` API when I started using it. Let me address these in a separate PR?

csilvestrini replied:
Separate PR SGTM. 

I don't really understand the type parameter issue you're describing. Does this not work?

```
inline fun getOrElse(onFailure: (Outcome<T>.Failure) -> T): T = when(this) {
    is Success -> value
    is Failure -> onFailure(this)
}

val x = Outcome.Failure<Int>("error")
x.getOrElse { 7 }
x.getOrElse { throw IllegalArgumentException(it.reason) }
x.getOrElse { return it }
```

bgogul replied:
`getOrElse` is inline. Therefore `return it` exits the function. 

```
fun foo(): Outcome<R> {
  val x = Outcome<T>.getOrElse { 
     // type mismatch as `it` is Outcome<T>, but foo() returns Outcome<R>
      return it 
   }
}
```

csilvestrini replied:
Ahh, thanks, now I understand what you meant. That's a shame :-( 

I can't think of anything elegant...

bgogul replied:
:(

cromwellian replied:
This is giving me strong "Option/Either" monad vibes, e.g. Either<Success, Failure> and getOrElse = fmap :)

bgogul replied:
Precisely. :) I went down the route of `getOrElse` as it was analogous to Kotlin's `Result` type.

csilvestrini commented on java/arcs/core/analysis/EntityTypeUnion.kt:
```diff
@@ -0,0 +1,70 @@
+): Outcome<Map<FieldName, FieldType>> {
+    val result = mutableMapOf<FieldName, FieldType>()
+    result.putAll(this)
+    other.forEach { entry ->
```
You can destructure the entry and use clearer names:

```
other.forEach { (name, type) ->
```

csilvestrini commented on javatests/arcs/core/analysis/EntityTypeUnionTest.kt:
```diff
@@ -0,0 +1,100 @@
+        val textSchema = Schema(setOf(SchemaName("Example")), textField, "")
+        val numberSchema = Schema(setOf(SchemaName("Example")), numberField, "")
+        val result = textSchema.union(numberSchema).getOrNull()
+        result?.let {
```
This will skip running the assertions below if result is null, which is not what you want to happen.

You have a few options here:

```
val result = textSchema.union(numberSchema).getOrNull()!!
val result = requireNotNull(textSchema.union(numberSchema).getOrNull())
val result = textSchema.union(numberSchema).getOrThrow()
```

(`getOrThrow` you'd have to write. In Rust they call that function `unwrap`, but I don't know if that's a common enough term that everyone would understand it)

bgogul replied:
I meant to add a `assertThat(result).isNotNull()` before this call, but seems to have missed it in this case. Added it now.

csilvestrini replied:
This construction is still pretty dangerous, since forgetting the `isNotNull` call can cause it to completely skip all of the assertions, whereas with the other options I suggested, forgetting to use them causes a compile error (nullability).

I'd recommend `!!` as the very easiest option.

bgogul replied:
Ah, makes sense. Used `requireNotNull` as it is more readable.

csilvestrini replied:
Sweet thanks! Now there's no need for the `result?.let` at all

bgogul replied:
Good point! removed `?.`.

bgogul commented:
@csilvestrini Thanks for the review. I addressed your comments. I also had a question about your suggestion for `getOrElse`. Please take a look when you get a chance.

csilvestrini commented on java/arcs/core/analysis/EntityTypeUnion.kt:
```diff
@@ -0,0 +1,70 @@
+ * If the maps have common [FieldName] entries, the union succeeds if and only if the corresponding
+ * [FieldType] values are the same.
+ */
+private fun Map<FieldName, FieldType>.unionFields(
```
This can be an infix fun too

bgogul replied:
It feels a bit weird to have `a unionFields b`. Also `unionFields` is local to this file. So, I will leave this as is.

csilvestrini commented on java/arcs/core/analysis/EntityTypeUnion.kt:
```diff
@@ -0,0 +1,70 @@
+ * Computes the union of [SchemaField] instances. Returns [Outcome.Failure] if the union
+ * results in any incompatibility. e.g., incompatible [FieldType] with the same name.
+ */
+private fun SchemaFields.union(other: SchemaFields): Outcome<SchemaFields> {
```
And this

bgogul replied:
Done.

csilvestrini commented on java/arcs/core/analysis/EntityTypeUnion.kt:
```diff
@@ -0,0 +1,70 @@
+ * is not possible as the inputs are incompatible.
+ */
+infix fun Schema.union(other: Schema): Outcome<Schema> {
+    val newNames = names.union(other.names)
```
`names union other.names`

bgogul replied:
Done.

csilvestrini commented on java/arcs/core/analysis/EntityTypeUnion.kt:
```diff
@@ -0,0 +1,70 @@
+ */
+infix fun Schema.union(other: Schema): Outcome<Schema> {
+    val newNames = names.union(other.names)
+    val newFields = fields.union(other.fields).getOrElse { return Outcome.Failure(it) }
```
`(fields union other.fields)`

bgogul replied:
Done.

csilvestrini commented on java/arcs/core/analysis/EntityTypeUnion.kt:
```diff
@@ -0,0 +1,70 @@
+ * results in any incompatibility. e.g., incompatible [FieldType] with the same name.
+ */
+private fun SchemaFields.union(other: SchemaFields): Outcome<SchemaFields> {
+    val newSingletons = singletons.unionFields(other.singletons)
```
`singletons unionFields other.singletons`

csilvestrini commented on javatests/arcs/core/analysis/EntityTypeUnionTest.kt:
```diff
@@ -0,0 +1,102 @@
+        val result = (textEntity union numberEntity).getOrNull()
+        assertThat(result).isNotNull()
+        result?.let {
+            assertThat(it.entitySchema).isEqualTo(numberSchema.union(textSchema).getOrNull())
```
`numberSchema union textSchema`

Cypher1 commented on java/arcs/core/analysis/EntityTypeUnion.kt:
```diff
@@ -0,0 +1,70 @@
+    result.putAll(this)
+    other.forEach { (name, type) ->
+        val existing = this.get(name)
+        if (existing != null && type != existing) {
```
I think this should actually be unioning the types of the fields.
Often this is as simple as `==` (for primitive types) but for non-primitives (e.g. references, refined types etc.) there's more to do.

Don't mind if this behavior change is implemented as a stub or if you just put a todo here to follow up on it.

bgogul replied:
Yep. It needs to be changed for non-primitive types. I swear I had a TODO, but should have gotten dropped during some refactoring. :) I have added a TODO here.

Cypher1 commented on javatests/arcs/core/analysis/EntityTypeUnionTest.kt:
```diff
@@ -0,0 +1,102 @@
+
+    @Test
+    fun schemaUnionMergesNames() {
+        val emptySchemaFields = SchemaFields(emptyMap(), emptyMap())
```
I have a feeling we'll need this type a lot. Might be worth defining it centrally somewhere so that we don't have to rewrite every creation of it if we change the representation.

e.g. something like `SchemaFields.getTopSchema()`

bgogul replied:
I agree, we might need something like this. Will add it in a separate PR as the need arises. 

BTW, why `TopSchema`? Do you mean `Top` in the sense of the top value in an abstract domain?

Cypher1 commented on javatests/arcs/core/analysis/EntityTypeUnionTest.kt:
```diff
@@ -0,0 +1,101 @@
+        val textEntity = EntityType(textSchema)
+        val numberEntity = EntityType(numberSchema)
+        val result = requireNotNull((textEntity union numberEntity).getOrNull())
+        result?.let {
```
This `?.` shouldn't be needed with `requireNotNull` right?

bgogul replied:
You are right. Removed `?.`

bgogul commented:
/gcbrun

bgogul commented:
> Might be good later to generalize Outcome for reuse

Agreed. Will keep an eye on this.
